### ISSUE TITLE
void_checks is too strict

### ISSUE THREAD

**Issue by MichaelRFairhurst** [dart-lang/linter#1107]
I don't think it should cover this case:

```
void Function(T) f;
T f2(T) { ... }

f = f2;
```

It is very common that functions which have side-effects also return a value, because (for instance) otherwise that value is completely lost:

```
list.removeAt(1); // does not return void, returns the removed item
```

That makes it impossible to assign it to places purely for the side-effect with this lint turned on.

```
Stream<int> f;
f.listen(list.removeAt); // LINT
```

This, to me, is very different than something like:

```
Completer<void>().complete(list.removeAt);
```

In the former case, you are passing a value in a place where it is used, and a part of that value is discarded but there's a lot of indication you knew what you were doing (you passed a unary int function in where a unary int function was expected). And the code in that form is the cleanest it can be.

In the latter case, the entire value you provided is guaranteed (ish*) to be discarded, and there's no indication that there's any relationship between the value asked and the value provided (except in that everything is related to `void`). And the code would be cleaner as:

```
// if "list.removeAt" is an expression with side-effects, it should go here. Otherwise it should be dropped.
Completer<void>().complete(null);
```

I hate to present this contrarian case because its an edge case that was probably fun & challenging to cover (nice work :)) but I unfortunately don't think it's a service to users and should be moved to a different lint if there's disagreement about that, probably.

**Comment by MichaelRFairhurst** [comment]
I can provide numbers on this: I have a large codebase broken into build targets with bazel, and about a fifth of our build targets either are broken from this lint or depend on targets that are broken from this lint.

Removing this case (I think I did so correctly...!) cut it down to about half (so about 10%). Three errors are the cause of 9.5% of that.

I didn't see how much about how the other 10% breaks down (could also be that 9.5% is caused by 3 cases). So the numbers aren't completely fair.

In any case, those are the numbers that I have that I can share :)

**Comment by eernstg** [comment]
*tl;dr Agreed, it is too strict to flag `f = f2`  rd;lt*

**Voidness preservation** didn't make it into Dart 2 as a part of the type system, but we may still introduce it as a lint. Maybe `void_checks` is that lint, already?

In the discussions about voidness preservation we always put the emphasis on helping developers avoid using a value which is obtained from the evaluation of an expression whose static type is `void`, and we never wanted to support the converse (that is, we did not aim to help developers avoid forgetting a useful value by passing it into a `void` typed target)<sup>1</sup>. In short "**a void value _should_ be discarded, and a non-void value _can_ be discarded**".

This means that it is perfectly in line with the expected voidness preservation rules to allow `f = f2` in the [original example](https://github.com/dart-lang/linter/issues/1107#issue-347486207): It is a higher-order variant of the situation where a non-void value is discarded, because invocations of `f` will require the returned value (whose type is `void`) to be discarded, even though the returned value from an invocation of the same function via `f2` had a non-void type.

Footnote 1: We have an **exception**: It was/is a warning/error if a non-void value is discarded by `return e` in a block-bodied function whose return type is `void`. So that's a rule of the form "**a non-void value _should not_ be discarded**". But that's a special case that only applies at `return` statements, and everything about voidness preservation is concerned with the opposite situation, that is, "a void value _should_ be discarded".

**Comment by MichaelRFairhurst** [comment]
Voidness preservation is not handled by this lint but could be. I wasn't a big fan of the name 'void_checks', but it works well for handling both voidness preservation _and_ assign-to-void smells. And I don't see a great reason why they should be two separate lints.

Its also notable that so far that none of the 3 cases of the lint I've fixed so far have been a problem, indicating that the theory behind why its valid to throw things away is not in stark contrast to reality.

But I still in many ways want to support void_checks for the following use case:

```
  // if users *want* to make a StreamController that cannot take any events but Null, I support that
  // even if I think they're fighting bugs that aren't very likely or strictly related to soundness.
  StreamController<Null> ctrl = new StreamController();
  ctrl.stream.listen((_) {
    // however, here _ has type Null.
  });
```

So I like the option of making void some loose definition of unassignable _and_ unusable. That way users don't have to bother ever using Null, and can't make a mistake about using Null when void has subtle reasons of being preferable.

**Comment by MichaelRFairhurst** [comment]
@a14n for thoughts

**Comment by a14n** [comment]
IIUC it's quite similar to the line with `f3` in this test:

https://github.com/dart-lang/linter/blob/a088573c6628935a634d642d0863b7fd58b106a6/test/rules/void_checks.dart#L134-L149

As the usage of `f3` is OK we should change the lint to cover your cases.

**Comment by MichaelRFairhurst** [comment]
Closing this, but we may want to open new ones for the following cases:

- voidness preservation
- `void async()` which seems to be commonly misused
- casting a `Future` to a `void` more strictly (for instance, `Future Function()` to `void Function()`) assuming that Futures should be thrown away much less often.
